examples/tv-app/tv-common/clusters/messages/MessagesManager.cpp

```diff
@@ -18,64 +18,110 @@
-        // TODO: Convert responses to Optional<chip::app::DataModel::List<const
-        // chip::app::Clusters::Messages::Structs::MessageResponseOptionStruct::Type>> message.responses = responses;
-    };
-
-    mMessages.push_back(message);
+    ChipLogProgress(Zcl, "HandlePresentMessagesRequest message:%s", std::string(messageText.data(), messageText.size()).c_str());
+
+    VerifyOrReturn(messageId.size() == CachedMessage::kMessageIdLength,
+                   ChipLogProgress(Zcl, "HandlePresentMessagesRequest invalid message id length"));
+
+    auto cachedMessage = CachedMessage(messageId, priority, messageControl, startTime, duration,
+                                       std::string(messageText.data(), messageText.size()));
+    if (responses.HasValue())
+    {
+        size_t size    = 0;
+        CHIP_ERROR err = responses.Value().ComputeSize(&size);
+        VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogProgress(Zcl, "HandlePresentMessagesRequest size check failed"));
+
+        VerifyOrReturn(size <= CachedMessage::kMaxOptionCount,
+                       ChipLogProgress(Zcl, "HandlePresentMessagesRequest too many options"));
+
+        auto iter = responses.Value().begin();
+        while (iter.Next())
+        {
+            auto & response = iter.GetValue();
+
+            VerifyOrReturn(response.messageResponseID.HasValue() && response.label.HasValue(),
+                           ChipLogProgress(Zcl, "HandlePresentMessagesRequest missing response id or label"));
+
+            VerifyOrReturn(response.messageResponseID.Value() >= CachedMessageOption::kResponseIdMin,
+                           ChipLogProgress(Zcl, "HandlePresentMessagesRequest responseID value check failed"));
+
+            VerifyOrReturn(response.label.Value().size() <= CachedMessageOption::kLabelMaxLength,
+                           ChipLogProgress(Zcl, "HandlePresentMessagesRequest label length check failed"));
+
+            CachedMessageOption option(response.messageResponseID.Value(),
+                                       std::string(response.label.Value().data(), response.label.Value().size()));
+
+            cachedMessage.AddOption(option);
+        }
+        VerifyOrReturn(iter.GetStatus() == CHIP_NO_ERROR, ChipLogProgress(Zcl, "HandlePresentMessagesRequest TLV parsing error"));
+    }
+
+    mCachedMessages.push_back(cachedMessage);
+
     // Add your code to present Message
+    ChipLogProgress(Zcl, "HandlePresentMessagesRequest complete");
 }
 
-void MessagesManager::HandleCancelMessagesRequest(const chip::app::DataModel::DecodableList<chip::ByteSpan> & messageIds)
+void MessagesManager::HandleCancelMessagesRequest(const DataModel::DecodableList<ByteSpan> & messageIds)
 {
-    // TODO: Cancel Message
+    auto iter = messageIds.begin();
+    while (iter.Next())
+    {
+        auto & id = iter.GetValue();
+
+        mCachedMessages.remove_if([id](CachedMessage & entry) {
+            if (entry.mMessageId.data_equal(id))
```

Comment:
the pattern of `if (foo) return true; return false` is odd. Use `return foo`. In this case:

```cpp
return entry.mMessageId.data_equal(id);
```

Reply:
addressed